extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufReader.java

```diff
@@ -0,0 +1,226 @@
+    return resourceMetricsList.stream()
+            .flatMap(
+            resourceMetrics -> {
+              List<KeyValue> resourceAttributes = resourceMetrics.getResource().getAttributesList();
+              return resourceMetrics.getInstrumentationLibraryMetricsList()
+                      .stream()
+                      .flatMap(libraryMetrics -> libraryMetrics.getMetricsList()
+                              .stream()
+                              .map(metric -> parseMetric(metric, resourceAttributes))
+                              .flatMap(List::stream)
+                      );
+            }
+    ).collect(Collectors.toList());
+  }
+
+  private List<InputRow> parseMetric(Metric metric, List<KeyValue> resourceAttributes)
+  {
+
+    List<String> schemaDimensions = dimensionsSpec.getDimensionNames();
+    if (!schemaDimensions.isEmpty()) {
+      return createRows(metric, resourceAttributes, schemaDimensions::contains);
+    } else {
+      return createRows(metric, resourceAttributes, att -> !dimensionsSpec.getDimensionExclusions().contains(att));
+    }
+  }
+
+  private List<InputRow> createRows(Metric metric,
+                                    List<KeyValue> resourceAttributes,
+                                    Function<String, Boolean> attributeMembershipFn)
+  {
+
+    List<NumberDataPoint> dataPoints = getDataPoints(metric);
+    List<InputRow> rows = new ArrayList<>();
+
+    for (NumberDataPoint dataPoint : dataPoints) {
+
+      final long timeUnixMilli = TimeUnit.NANOSECONDS.toMillis(dataPoint.getTimeUnixNano());
+      final List<KeyValue> metricAttributes = dataPoint.getAttributesList();
+      final Number value = getValue(dataPoint);
+
+      final int capacity = resourceAttributes.size()
+              + metricAttributes.size()
+              + 2; // metric name + value columns
+
+      Map<String, Object> event = Maps.newHashMapWithExpectedSize(capacity);
+      event.put(metricDimension, metric.getName());
+      event.put(VALUE_COLUMN, value);
+
+      List<String> dimensions = new ArrayList<>();
+      dimensions.add(metricDimension);
+      dimensions.add(VALUE_COLUMN);
+
+      resourceAttributes.stream()
+              .filter(ra -> attributeMembershipFn.apply(this.resourceAttributePrefix + ra.getKey()))
+              .forEach(ra -> {
+                dimensions.add(this.resourceAttributePrefix + ra.getKey());
+                event.put(this.resourceAttributePrefix + ra.getKey(), ra.getValue().getStringValue());
+              });
+
+      metricAttributes.stream()
+              .filter(ma -> attributeMembershipFn.apply(this.metricAttributePrefix + ma.getKey()))
+              .forEach(ma -> {
+                dimensions.add(this.metricAttributePrefix + ma.getKey());
+                event.put(this.metricAttributePrefix + ma.getKey(), ma.getValue().getStringValue());
+              });
+
+      rows.add(new MapBasedInputRow(timeUnixMilli, dimensions, event));
```

Comment:
It's not clear why we need to maintain a separate `List` for the dimension keys vs. just taking `event.keySet()` and converting it to a list.  If you do this you can remove the `dimensions` list and all the code that updates both it and the `event` hashmap.
```suggestion
      rows.add(new MapBasedInputRow(timeUnixMilli, new ArrayList<>(event.keySet()), event));
```

Reply:
I agree. I've made a change that reflects that. I noticed, though, that logically the opencensus inputformat will set the dimensions field to be the whole `SchemaDimensions.getDimensionNames()` list for all rows if it is not null, so I made a change to reflect that.

Reply:
the InputFormat is responsible for applying the schema defined in the input format (i.e. filtering out the dimensions that are not requested, and apply dimension exclusions) so we can't just take the event's dimensions, instead the row need to contain the effective dimensions that the user has requested.

Reply:
Makes sense, and I think the latest code reflects that.  In the original code, the dimension list was redundant with the `event.keySet()`.
> I noticed, though, that logically the opencensus inputformat will set the dimensions field to be the whole SchemaDimensions.getDimensionNames() list for all rows if it is not null, so I made a change to reflect that.

Reply:
@xvrl How should we group datapoints' attributes so that all rows that come from one group have the same dimension when there is no configured `schemaDimension` and we are calculating "effective dimensions"? Should all rows derived from a single `MetricData` have the same dimensions, or would it make sense to have all metrics under the same `Resource` share the same dimensions, or is it okay to have each datapoint's attributes define the "effective dimensions"? 

I'm not sure what the downstream correctness/performance implications are of the current implementation, but if we wanted all datapoints under the same `MetricData`, `Resource` or `Metric` to have a common dimension list we'd need to break up the row creation into two steps. 1) we gather the attributes of a group of datapoints into a set and 2) we remove exclusions from that set and then create the actual rows that come from the datapoints. Since splitting the one step into 2 adds extra complexity I was hoping it could be avoided.

```java
  List<Map<String,Object>> rowsData = metricsData.getResourceMetricsList()
          .stream()
          .flatMap(resourceMetrics -> {
            Map<String, Object> resourceData = resourceMetrics.getResource()
                    .getAttributesList()
                    .stream()
                    .collect(Collectors.toMap(KeyValue::getKey, KeyValue::getValue));

            Stream<Map<String, Object>> data = resourceMetrics.getInstrumentationLibraryMetricsList()
                    .stream()
                    .flatMap(libraryMetrics -> libraryMetrics.getMetricsList()
                            .stream()
                            .flatMap(metric -> parseMetric(metric, resourceData).stream()));
            return data;
          })
          .collect(Collectors.toList());

  return createRows(rowsData);

```

Reply:
when there is not configured schema, we want all the labels in the data minus the excluded dimensions. So, each datapoint defines the dimensions we have, and we remove what the user asked to remove. This allows someone to do schema-less ingestion in druid where you don't have to define a schema at all, and all the columns automatically get picked up. So all the resource labels should be present, and all the metric labels should be present. Each row could have different labels, since each could be a different metric.
This already works with the opencensus inputformat AFAIK, so we should follow the approach we have there already.